pkg/plugins/tainttoleration/taint_toleration.go

```diff
@@ -38,80 +44,150 @@ func (pl *TaintToleration) Description() string {
 	for _, toleration := range placement.Spec.Tolerations {
 		if len(toleration.Key) == 0 && toleration.Operator != clusterapiv1beta1.TolerationOpExists {
-			return nil, errors.New("If the key is empty, operator must be Exists.\n")
+			return plugins.PluginFilterResult{
+				Err: errors.New("If the key is empty, operator must be Exists.\n"),
+			}
 		}
 		if toleration.Operator == clusterapiv1beta1.TolerationOpExists && len(toleration.Value) > 0 {
-			return nil, errors.New("If the operator is Exists, the value should be empty.\n")
+			return plugins.PluginFilterResult{
+				Err: errors.New("If the operator is Exists, the value should be empty.\n"),
+			}
 		}
 	}
 
-	existingDecisions := getDecisions(pl.handle, placement)
+	decisionClusterNames := getDecisionClusterNames(pl.handle, placement)
 
 	// filter the clusters
 	matched := []*clusterapiv1.ManagedCluster{}
 	for _, cluster := range clusters {
-		if isClusterTolerated(cluster, placement.Spec.Tolerations, existingDecisions.Has(cluster.Name)) {
+		if tolerated, _ := isClusterTolerated(cluster, placement.Spec.Tolerations, decisionClusterNames.Has(cluster.Name)); tolerated {
 			matched = append(matched, cluster)
 		}
 	}
 
-	return matched, nil
+	return plugins.PluginFilterResult{
+		Filtered: matched,
+	}
+}
+
+func (pl *TaintToleration) RequeueAfter(ctx context.Context, placement *clusterapiv1beta1.Placement) plugins.PluginRequeueResult {
+	// get exist decisions clusters
+	decisionClusterNames, decisionClusters := getDecisionClusters(pl.handle, placement)
+	if decisionClusterNames == nil || decisionClusters == nil {
+		return plugins.PluginRequeueResult{}
+	}
+
+	requeueList := []*plugins.PluginRequeueResult{}
+	// filter and record pluginRequeueResults
+	for _, cluster := range decisionClusters {
+		tolerated, requeue := isClusterTolerated(cluster, placement.Spec.Tolerations, decisionClusterNames.Has(cluster.Name))
+		if tolerated && requeue != nil {
+			requeueList = append(requeueList, requeue)
+		}
+	}
+
+	requeueResult := minRequeueTime(requeueList)
+	if requeueResult == nil {
+		return plugins.PluginRequeueResult{}
+	}
+
+	return *requeueResult
 }
 
 // isClusterTolerated returns true if a cluster is tolerated by the given toleration array
-func isClusterTolerated(cluster *clusterapiv1.ManagedCluster, tolerations []clusterapiv1beta1.Toleration, inDecision bool) bool {
+func isClusterTolerated(cluster *clusterapiv1.ManagedCluster, tolerations []clusterapiv1beta1.Toleration, inDecision bool) (bool, *plugins.PluginRequeueResult) {
+	requeueList := []*plugins.PluginRequeueResult{}
 	for _, taint := range cluster.Spec.Taints {
-		if !isTaintTolerated(taint, tolerations, inDecision) {
-			return false
+		tolerated, requeue := isTaintTolerated(taint, tolerations, inDecision)
+		if !tolerated {
+			return false, nil
+		}
+		if requeue != nil {
+			requeueList = append(requeueList, requeue)
 		}
 	}
-	return true
+
+	return true, minRequeueTime(requeueList)
```

Comment:
nit: we do not need a separated sorting here. you could know the min requeue during the for loop. This could reduce some computation cost. But it can be a followup

Reply:
fixed.